virtual-display-driver-cli/src/client.rs

```diff
@@ -1,110 +1,139 @@
+use eyre::Context as _;
+use joinery::Joinable as _;
 use win_pipes::{NamedPipeClientReader, NamedPipeClientWriter};
 
 pub struct Client {
-    reader: NamedPipeClientReader,
     writer: NamedPipeClientWriter,
+    state: Vec<Monitor>,
 }
 
 impl Client {
     pub fn connect() -> eyre::Result<Self> {
-        let (reader, writer) = win_pipes::NamedPipeClientOptions::new("virtualdisplaydriver")
-            .wait()
-            .access_duplex()
-            .mode_message()
-            .create()?;
-
-        Ok(Self { reader, writer })
+        let (mut reader, mut writer) =
+            win_pipes::NamedPipeClientOptions::new("virtualdisplaydriver")
+                .wait()
+                .access_duplex()
+                .mode_message()
+                .create()
+                .context("Failed to connect to Virtual Display Driver; please ensure the driver is installed and working. Other program using the driver must also be closed, such as the Virtual Display Driver Control app.")?;
+
+        send_command(&mut writer, &driver_ipc::Command::RequestState)?;
+        let state = receive_command(&mut reader)?;
+        let driver_ipc::Command::ReplyState(state) = state else {
+            eyre::bail!("received unexpected reply from driver pipe");
+        };
+
+        Ok(Self { writer, state })
     }
 
-    pub fn list(&mut self) -> eyre::Result<Vec<Monitor>> {
-        self.send_command(&driver_ipc::Command::RequestState)?;
-        let response = self.receive_command()?;
-
-        match response {
-            driver_ipc::Command::ReplyState(state) => Ok(state),
-            _ => eyre::bail!("received unexpected reply from driver pipe"),
-        }
+    pub fn monitors(&self) -> &[Monitor] {
+        &self.state
     }
 
     pub fn get(&mut self, id: driver_ipc::Id) -> eyre::Result<Monitor> {
-        let monitors = self.list()?;
-        let monitor = monitors.into_iter().find(|monitor| monitor.id == id);
+        let monitor = self.state.iter().find(|monitor| monitor.id == id);
 
         match monitor {
-            Some(monitor) => Ok(monitor),
-            None => eyre::bail!("no virtual monitor with ID {} found", id),
+            Some(monitor) => Ok(monitor.clone()),
+            None => eyre::bail!("virtual monitor with ID {} not found", id),
         }
     }
 
     pub fn notify(&mut self, monitors: Vec<driver_ipc::Monitor>) -> eyre::Result<()> {
         let command = driver_ipc::Command::DriverNotify(monitors);
 
-        self.send_command(&command)?;
+        send_command(&mut self.writer, &command)?;
 
         Ok(())
     }
 
+    pub fn validate_has_ids(&self, ids: &[driver_ipc::Id]) -> eyre::Result<()> {
+        let ids = ids.iter().copied().collect::<BTreeSet<_>>();
+        let existing_ids = self
+            .state
+            .iter()
+            .map(|monitor| monitor.id)
+            .collect::<BTreeSet<_>>();
+        let missing_ids = ids.difference(&existing_ids).copied().collect::<Vec<_>>();
```

Comment:
Implementation note: I used `BTreeSet`s here so that the order that comes out of `missing_ids` gets sorted. Although it may be different than the order the user typed in the IDs (e.g. for `remove`), I felt that this was a good compromise between making this function simple and making the output human-friendly.

Reply:
I like the simplicity of this function. The ordering of the reported missing IDs isn't a huge deal, as any users (to the best of my knowledge) shouldn't be relying on that anyways.